gems/aws-sdk-core/lib/aws-sdk-core/profile/profile_file_parser.rb

```diff
@@ -0,0 +1,112 @@
+# frozen_string_literal: true
+
+module AWS::SDK::Core
+  # @api private
+  class ProfileFileParser
+    def initialize(profile_file_contents)
+      @profile_file_contents = profile_file_contents
+      @current_line_number = 0
+    end
+
+    def parse
+      initialize_profiles
+      @profiles
+    end
+
+    private
+
+    def initialize_profiles
+      @profiles = {}
+
+      @profile_file_contents.split(/[\r\n]+/).each do |line|
+        @current_line_number += 1
+        next if ProfileFileUtils.empty_line?(line) ||
+                ProfileFileUtils.comment_line?(line)
+
+        if ProfileFileUtils.profile_line?(line)
+          read_profile_line(line)
+        elsif ProfileFileUtils.property_continuation_line?(line)
+          read_property_continuation_line(line)
+        else
+          read_property_definition_line(line)
+        end
+      end
+    end
+
+    def read_profile_line(line)
+      line_without_comments = remove_trailing_comments(line, %w[# ;])
+      line_without_whitespace =
+        ProfileFileUtils.trim_whitespace(line_without_comments)
+
+      unless line_without_whitespace[-1] == ']'
+        raise ArgumentError,
```

Comment:
ArgumentError doesn't really feel like the right error to be raising for these. I'd lean towards creating an `InvalidProfileError` or something like that.

Reply:
I can make a profile error class